### DISCUSSION TITLE
SMARTS doesn't match where it should?

### DISCUSSION THREAD

**Discussion by boglet** [rdkit/rdkit#6610] [Q&A] [answered]
I freely admit to not being the best composer of SMARTS queries so this could just be me!

I have a SMARTS to identify 1,4-disubstituted benzene rings.  I have tried a couple of different versions and have had no luck in getting it to work.  The first one:
`c(-[H])c(-[H])c(-[!H])c(-[H])c(-[H])c(-[!H])`
doesn't match 
`C1C(CC)=CC=C(OC)C=1 `
or 
`c1c(CC)ccc(OC)c1`

Neither does the second one:
`[#6](-[#1]):[#6](-[#1]):[#6](-[!#1]):[#6](-[#1]):[#6](-[#1]):[#6]-[!#1]`

However, both of these match when I use the Indigo toolkit from EPAM. If anything, I suspect that the SMARTS is not specific enough and may hit fused rings too.

Is it me or is it a bug?

**Comment by bp-kelley** [comment, marked as answer]
What you are seeing is a "feature".  The SMILES line notation was designed to be very compact.  So compact, that in most cases hydrogens are not explicit.  You'll note that in your molecule:

```
C1C(CC)=CC=C(OC)C=1
```

you don't write hydrogens at all, however in your smarts you do:

```
c(-[H])c(-[H])c(-[!H])c(-[H])c(-[H])c(-[!H])
```

So this happens:

```
>>> sma = Chem.MolFromSmarts("c(-[H])c(-[H])c(-[!H])c(-[H])c(-[H])c(-[!H])")
>>> m = Chem.MolFromSmiles("C1C(CC)=CC=C(OC)C=1")
>>> m.HasSubstructMatch(sma)
False
```

But that is because the hydrogens are stored implicitly and you requested explicit hydrogens.  Now you can change the implicit to explicit:

```
>>> Chem.AddHs(m).HasSubstructMatch(sma)
True
```

But in general, you don't do this.  You simply use the hydrogens as atom properties in the smarts pattern, i.e. [cH] which is an aromatic carbon with one H, [cH2] an aromatic carbon with 2 Hs.

```
sma = Chem.MolFromSmarts("[cH][cH]c(-[!H])[cH][cH]c(-[!H])")
>>> m.HasSubstructMatch(sma)
False
```

And now we can simply use wildcards for the non-hs since hydrogens are implicit:
```
sma = Chem.MolFromSmarts("[cH][cH]c(-*)[cH][cH]c(-*)")
```

You could also use the MergeQueryHs which convert explicit to implicit in the query but might create a slightly different meaning than you intend

```
>>> sma = Chem.MolFromSmarts("c(-[H])c(-[H])c(-[!H])c(-[H])c(-[H])c(-[!H])")
>>> sma = Chem.MergeQueryHs(sma)
>>> print(Chem.MolToSmarts(sma))
'[c&!H0][c&!H0]c(-[!H1])[c&!H0][c&!H0]c-[!H1]'
```

 Hopefully this whirlwind tour helps.

**Comment by boglet** [comment]
Thanks so much for that detailed explanation. I understand how this works now in RDkit. Your SMARTS works beautifully for my use case, so a big thank you!

I feel that explicit or implicit hydrogens shouldn't affect the result of a substructure (or SMARTS) match. The chemical sense of the molecule or the query should be independent of the chosen representation. After all, if I represent the SMILES in aromatic or Kekule form, it will still be found by an aromatic SMARTS. This seems like a pretty big inconsistency (I know you mentioned that it is a "feature").

Am I alone in this, or do people feel that this behaviour is something that should be changed?
